### ISSUE TITLE
HTTP/1.0 reply without Content-Length

### ISSUE THREAD

**Issue by milabs** [tempesta-tech/tempesta#497]
There's a problem when server replies to `HTTP/1.0` request with response without `Content-Length` specified:

**request**
~~~
[   44.496528] [tempesta]   parse 62 client data bytes (GET /forum.php HTTP/1.0
               Host: 192.168.88.220:3128
               User-Agent) on req=ffff88007b5ab020
~~~

**reply**
~~~
[   44.888615] [tempesta]   parse 74 server data bytes (HTTP/1.1 200 OK
               Server: Tengine/1.3.0
               Date: Sun, 22 May 2016 21:35:51 GM) on resp=ffff88007b5a8020
~~~

~~~
[   44.921069] [tempesta]   parse 1366 server data bytes (T
               Content-Type: text/html; charset=gbk
               Connection: close
               X-Powered-By: PHP/5.3.14
               Set-Cookie: MECz_2132_saltkey=tzmzMLah; expires=Tue, 21-Jun-2016 21:35:51 GMT; path=/; httponly
               Set-Cookie: MECz_2132_lastvisit=1463949351; expires=Tue, 21-Jun-2016 21:35:51 GMT; path=/
               Set-Cookie: MECz_2132_sid=LuYGf1; expires=Mon, 23-May-2016 21:35:51 GMT; path=/
               Set-Cookie: MECz_2132_lastact=1463952951%09forum.php%09; expires=Mon, 23-May-2016 21:35:51 GMT; path=/
               
               <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
               <html xmlns="http://www.w3.org/1999/xhtml">
               <head>
               <meta http-equiv="Content-Type" content="text/html; charset=gbk" />
~~~

This occurs while running `ab` which uses `HTTP/1.0` for the request. Tempesta runs as a proxy for the `server 106.185.29.4:80`. So, the server does not give content length but closes the stream upon finishing the response. We must handle it properly as according to [RFC 2616 / 4.4](http://greenbytes.de/tech/webdav/rfc2616.html#message.length) resetting connection is the one of the possible signaling methods.

**Comment by krizhanovsky** [comment]
There is `conn_drop` callback called from `tfw_connection_drop()`. The hook is essentially `tfw_http_conn_drop()` which calls `tfw_http_conn_msg_free()`. The last one puts connections and frees the message, so we should resolve the task at `tfw_http_conn_drop()` while all connections and messages are alive.

A new HTTP parser function `tfw_http_parse_break()` or similar should be called from `tfw_http_conn_drop()` before `tfw_http_conn_msg_free()`. The function must check the parser status and return the value. It must clearly say whether the message is fully parsed (we're in correct body parsing state) or not.

Next `tfw_http_req_process()` and `tfw_http_resp_process()` must be split to 2 functions each: skb processing and calling HTTP parser and actual HTTP message logic (e.g. https://github.com/natsys/tempesta/blob/master/tempesta_fw/http.c#L798 - https://github.com/natsys/tempesta/blob/master/tempesta_fw/http.c#L909 for request processing). The HTTP message logic functions must be called from `tfw_http_conn_drop()` like `tfw_http_msg_process()` does it.

**Comment by keshonok** [comment]
The relevant current RFC and section is [rfc7230 section-3.3.3](https://tools.ietf.org/html/rfc7230#section-3.3.3).

**Comment by krizhanovsky** [comment]
Please also consider this [blog post](https://www.mnot.net/blog/2011/07/11/what_proxies_must_do), "Detect Bad Framing", and especially [Response Splitting Attack](http://projects.webappsec.org/w/page/13246931/HTTP-Response-Splitting).

**Comment by keshonok** [comment]
Fixed via #545.
